doc/flow-definition.md

```diff
@@ -1,11 +1,308 @@
+
+All blocks of a particular type (within a single question) are executed at particular points in the request processing sequence i.e. all `on_response` blocks are executed before all `next_node_calculation` blocks. The order in which the blocks are defined only affects the order in which they are executed within the group of blocks of the same type i.e. when two `on_response` blocks are defined, the one defined first will be executed before the one defined second; however, even if a `next_node_calculation` block is defined before both of these `on_response` blocks, it will always be executed after both of them.
+
+The block types are executed in the following order:
+
+* [`precalculate`](#precalculatevariable_name-block)
+* [`on_response`](#on_responseblock)
+* [`next_node_calculation`](#next_node_calculationvariable_name-block)
+* [`validate`](#validatemessage_key-block)
+* [`next_node`](#next_nodeblock)
+* [`save_input_as`](#save_input_asvariable_name)
+* [`calculate`](#calculatevariable_name-block)
+
+Each of these block types and the point at which they are executed is explained in more detail below:
+
+##### `precalculate(variable_name, &block)`
+
+* These blocks were intended to be used to store state variables that are needed to render a question or outcome node template.
+* It's simplest to think about these blocks executing before the user response is parsed from the request path.
+* The block return value is stored on the state object as a state variable named `variable_name`.
+
+> Note that due to an oversight in the implementation of `Flow#process`, it's not currently possible to use a `precalculate` block in the first question.
+
+> The use of these blocks is deprecated and should never be necessary. Define methods on a `calculator` object instead.
+
+##### `on_response(&block)`
+
+* These blocks are intended to be used to store user responses on a `calculator` state variable. They are a [relatively new addition to the DSL][introduction-of-on-response-blocks].
+* These blocks are called  after the user response has been parsed from the request path and before any of the `next_node_calculation` blocks are executed.
+* The parsed response is passed to the block as the only argument and by convention is named `response`.
+* The block return value is not used and no state variable is stored.
+
+> The use of these blocks is encouraged; however, they should only ever be used to store a single, `calculator` state variable (in the first question definition); otherwise they should only be used to store user responses on that `calculator` object.
+
+##### `next_node_calculation(variable_name, &block)`
+
+* These blocks were intended to be used to store state variables that are needed in the `next_node` block.
```

Comment:
That's pretty much correct, although I think it would be clearer to call the "variable" on the calculator an "attribute" to distinguish it from the concept of a "state variable".

The `on_response` block type is something we have introduced, because we're trying to move to the situation where only a _single_ state variable (`calculator`) is *ever* set on the magic state object. Once we've done that for all flows, we plan to get rid of the magic state object altogether and use the calculator object instead.

I think this thinking is an example of something that would be better explained in person with the whole team. Once we've chatted about it, perhaps the "wiki elves" can improve what I've written here...?